tsdb/db.go

```diff
@@ -1161,6 +1176,17 @@ func (db *DB) Compact(ctx context.Context) (returnErr error) {
 		default:
 		}
 		if !db.head.compactable() {
+			if !db.compactStart.IsZero() {
+				db.compactStart = time.Time{}
+			}
+			break
+		}
+
+		if db.compactStart.IsZero() {
+			// Start counting for the offset.
+			db.compactStart = time.Now()
+		}
+		if time.Since(db.compactStart) < db.opts.CompactOffset {
```

Comment:
IIUC, If the offset in not enabled, `db.opts.CompactOffset` would be `0` and because `time.Since(db.compactStart)` would be `>=0` as you said, the comparison will yield false and the head compaction will run right away as expected.
I think there are some tests that implicitly rely on/verify that behavior, but I'll add an explicit test.